# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/wasm/WasmBaselineCompile.cpp
# Commit: 44b0c0bed8a6
# Full Hash: 44b0c0bed8a6b4debb950e7e1c69190883c65d60
# Author: Lars T Hansen <lhansen@mozilla.com>
# Date: 2021-10-18 16:09:29
# Regressor Bug: 1727084
# File Overlap Count: 1
# Description:
#   Bug 1727084 - Memory64 - Atomic-rmw baseline operations. r=yury
#   
#   Atomic ADD, SUB, AND, OR, XOR, XCHG, CMPXCHG, as well as WAIT and
#   NOTIFY for the baseline compiler.
#   
# ==============================================================================

diff -r fb02656bb788 -r 44b0c0bed8a6 js/src/wasm/WasmBaselineCompile.cpp
--- a/js/src/wasm/WasmBaselineCompile.cpp	Mon Oct 18 09:58:15 2021 +0000
+++ b/js/src/wasm/WasmBaselineCompile.cpp	Mon Oct 18 09:58:15 2021 +0000
@@ -291,6 +291,31 @@
 #endif
 }
 
+#ifdef JS_CODEGEN_X86
+void BaseCompiler::stashI64(RegPtr regForTls, RegI64 r) {
+  MOZ_ASSERT(sizeof(TlsData::baselineScratch) >= 8);
+  MOZ_ASSERT(regForTls != r.low && regForTls != r.high);
+  fr.loadTlsPtr(regForTls);
+  masm.store32(r.low, Address(regForTls, offsetof(TlsData, baselineScratch)));
+  masm.store32(r.high,
+               Address(regForTls, offsetof(TlsData, baselineScratch) + 4));
+}
+
+void BaseCompiler::unstashI64(RegPtr regForTls, RegI64 r) {
+  MOZ_ASSERT(sizeof(TlsData::baselineScratch) >= 8);
+  fr.loadTlsPtr(regForTls);
+  if (regForTls == r.low) {
+    masm.load32(Address(regForTls, offsetof(TlsData, baselineScratch) + 4),
+                r.high);
+    masm.load32(Address(regForTls, offsetof(TlsData, baselineScratch)), r.low);
+  } else {
+    masm.load32(Address(regForTls, offsetof(TlsData, baselineScratch)), r.low);
+    masm.load32(Address(regForTls, offsetof(TlsData, baselineScratch) + 4),
+                r.high);
+  }
+}
+#endif
+
 //////////////////////////////////////////////////////////////////////////////
 //
 // Function entry and exit